src/servers/src/mysql/helper.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::ops::ControlFlow;
+use std::time::Duration;
+
+use chrono::{NaiveDate, NaiveDateTime};
+use common_query::prelude::ScalarValue;
+use datatypes::prelude::ConcreteDataType;
+use datatypes::value::{self, Value};
+use opensrv_mysql::{ParamValue, ValueInner};
+use snafu::ResultExt;
+use sql::ast::{visit_expressions_mut, Expr, Value as ValueExpr, VisitMut};
+use sql::statements::statement::Statement;
+
+use crate::error::{self, Result};
+
+/// Returns the placeholder string "$i".
+pub fn format_placeholder(i: usize) -> String {
+    format!("${}", i)
+}
+
+/// Replace all the "?" placeholder into "$i" in SQL,
+/// returns the new SQL and the last placeholder index.
+pub fn replace_placeholder(query: &str) -> (String, usize) {
+    let mut query = query.to_string();
+    let mut index = 1;
+    let mut offset = 0;
+    while let Some(position) = query[offset..].find('?') {
+        let place_holder = format!("${}", index);
+        let position = position + offset;
+        query.replace_range(position..position + 1, &place_holder);
+        index += 1;
+        offset = position + 1;
+    }
+    (query, index)
```

Comment:
Seems the time and space complexity are both O(N^2)? Every time you find the next "?" and insert the "$x" in the middle of string. Can it rewrite like this:

```suggestion
    let query_parts = query.split('?');
    let mut index = 0;
    let query = query_parts.map(|x| {
        i += 1;
        format("{x} ${i}")
    }).join(" ");
    (query, index + 1)
```